src/main/java/com/es/phoneshop/model/product/ArrayListProductDao.java

```diff
@@ -59,57 +56,47 @@ public Product getProduct(Long id) {
     public List<Product> findProducts(String query, SortField sortField, SortType sortType) {
         readLock.lock();
         try {
+            String[] queryWords = (query != null && !query.equals("")) ? query.split("\\s+") : null;
             return products.stream()
-                    .filter(product ->
-                                    query == null ||
-                                    query.equals("") ||
-                                    search(query, product))
                     .filter(product -> product.getStock() > 0)
                     .filter(product -> product.getPrice() != null)
-                    .sorted((p1, p2) -> relevanceSort(query, p1, p2))
+                    .map(product -> Map.entry(product, getNumberOfCollisions(queryWords, product)))
+                    .filter(entry -> entry.getValue() > 0)
                     .sorted(getSortComparator(sortField, sortType))
+                    .map(Map.Entry::getKey)
                     .collect(Collectors.toList());
         } finally {
             readLock.unlock();
         }
     }
 
-    private boolean search(String query, Product product) {
-        return Arrays.stream(query.split("\\s+"))
-                .anyMatch(word -> product.getDescription().toLowerCase()
-                        .contains(word.toLowerCase()));
-    }
-
-
-    private int relevanceSort(String query, Product p1, Product p2) {
-        if (query == null || query.equals("")) return 0;
-        return Long.compare(getNumberOfCollisions(query, p2), getNumberOfCollisions(query, p1));
-    }
-
-    private long getNumberOfCollisions(String query, Product product) {
-        return Arrays.stream(query.split("\\s+"))
-                .filter(word -> product.getDescription().toLowerCase()
-                        .contains(word.toLowerCase()))
+    private long getNumberOfCollisions(String[] queryWords, Product product) {
+        if (queryWords == null) return 1;
+        return Arrays.stream(queryWords).filter(word -> product.getDescription().toLowerCase()
+                .contains(word.toLowerCase()))
                 .count();
     }
 
-    private Comparator<Product> getSortComparator(SortField field, SortType type) {
-        if (field == null) return (o1, o2) -> 0;
+    private Comparator<Map.Entry<Product, Long>> getSortComparator(SortField field, SortType type) {
+        if (field == null || type == null) return this::relevanceSort;
         return Comparator.comparing(
                 product -> {
-                    if (field == SortField.price) return product.getPrice();
-                    else return product.getDescription();
+                    if (field == SortField.price) return product.getKey().getPrice();
+                    else return product.getKey().getDescription();
                 },
                 getOrderTypeComparator(type)
```

Comment:
Won't comparator run this `if` (and decide on if he need to take price or description for comparison) every time it will need to compare two values?  No need to fix, just think it over.